dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/MsgPackStatefulSerializer.java

```diff
@@ -37,99 +44,192 @@
     }
   }
 
-  private MsgPackTraceBuffer newBuffer() {
-    MsgPackTraceBuffer buffer = new MsgPackTraceBuffer();
-    int traceBufferSize = traceBufferSize();
-    buffer.buffer = new ArrayBufferOutput(traceBufferSize);
-    return buffer;
+  @Override
+  public MsgPackTraceBuffer newBuffer() {
+    return new MsgPackTraceBuffer(new ArrayBufferOutput(bufferSize));
   }
 
-  private void updateTraceSize(int traceSize) {
+  private void updateTraceSizeEstimate(int traceSize) {
     traceSizeSum = (traceSizeSum - traceSizes[position] + traceSize);
     traceSizes[position] = traceSize;
     position = (position + 1) & (traceSizes.length - 1);
   }
 
-  private int traceBufferSize() {
-    // round up to next KB, assumes for now that there will be one trace per buffer
-    return ((lastTracesPerBuffer * traceSizeSum / TRACE_HISTORY_SIZE) + 1023) / 1024;
+  private int avgTraceSize() {
+    return traceSizeSum / TRACE_HISTORY_SIZE;
   }
 
-  public static class MsgPackTraceBuffer implements TraceBuffer {
+  static class MsgPackTraceBuffer implements TraceBuffer {
+
+    private static final AtomicInteger BUFFER_ID = new AtomicInteger(0);
 
-    private ArrayBufferOutput buffer;
+    private final ArrayBufferOutput buffer;
+    final int id;
     private int length;
     private int traceCount;
+    private int representativeCount;
+    private CountDownLatch flushLatch;
+
+    public MsgPackTraceBuffer(ArrayBufferOutput buffer) {
+      this.buffer = buffer;
+      this.id = BUFFER_ID.getAndIncrement();
+    }
 
     @Override
     public void writeTo(WritableByteChannel channel) throws IOException {
+      writeHeader(channel);
       int remaining = length;
       for (MessageBuffer messageBuffer : buffer.toBufferList()) {
         int size = messageBuffer.size();
-        ByteBuffer buffer =
-            size > remaining
-                ? messageBuffer.sliceAsByteBuffer(0, remaining)
-                : messageBuffer.sliceAsByteBuffer();
+        ByteBuffer buffer = messageBuffer.sliceAsByteBuffer(0, Math.min(size, remaining));
         while (buffer.hasRemaining()) {
-          channel.write(buffer);
+          remaining -= channel.write(buffer);
         }
-        remaining -= size;
       }
       assert remaining == 0;
     }
 
+    private void writeHeader(WritableByteChannel channel) throws IOException {
+      // inlines behaviour from MessagePacker.packArrayHeader
+      if (traceCount < (1 << 4)) {
+        ByteBuffer buffer = ByteBuffer.allocate(1);
+        buffer.put(0, (byte) (traceCount | FIXARRAY_PREFIX));
+        channel.write(buffer);
+      } else if (traceCount < (1 << 16)) {
+        ByteBuffer buffer = ByteBuffer.allocate(3);
+        buffer.put(0, ARRAY16);
+        buffer.putShort(1, (short) traceCount);
+        channel.write(buffer);
+      } else {
+        ByteBuffer buffer = ByteBuffer.allocate(5);
+        buffer.put(0, ARRAY32);
+        buffer.putInt(1, traceCount);
+        channel.write(buffer);
+      }
```

Comment:
Can this be combined with a call to `headerSize()`?

```suggestion
      int headerSize = headerSize();
      ByteBuffer buffer = ByteBuffer.allocate(headerSize);
      switch (headerSize) {
      if (traceCount < (1 << 4)) {
        case 1:
          buffer.put(0, (byte) (traceCount | FIXARRAY_PREFIX));
          break;
        case 3:
          buffer.put(0, ARRAY16);
          buffer.putShort(1, (short) traceCount);
          break;
        case 5:
          buffer.put(0, ARRAY32);
          buffer.putInt(1, traceCount);
          break;
        default:
          throw new IOException("Unsupported header size {}", headerSize);
      }
      channel.write(buffer);
```

Reply:
I'm not sure this is more succinct, whereas it's already inlining some logic from message pack so divergence _here_ may not be the primary concern. I will think about it.

Reply:
ok